widget/entry-with-completion.go

```diff
@@ -0,0 +1,156 @@
+package widget
+
+import (
+	"fmt"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/data/binding"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+// CompletionEntry is an Entry with options displayed in a PopUpMenu.
+type CompletionEntry struct {
+	widget.Entry
+	popupMenu     *widget.PopUp
+	navigableList *navigableList
+	items         []string
+	pause         bool
+}
+
+// NewCompletionEntry creates a new CompletionEntry which creates a popup menu that responds to keystrokes to navigate through the items without losing the editing ability of the text input.
+func NewCompletionEntry(options []string) *CompletionEntry {
+	c := &CompletionEntry{items: options}
+	c.ExtendBaseWidget(c)
+	return c
+}
+
+// ShowCompletion displays the completion menu
+func (c *CompletionEntry) ShowCompletion() {
+	if c.pause {
+		return
+	}
+
+	if c.navigableList == nil {
+		c.navigableList = newNavigableList(c.items, &c.Entry, c.setTextFromMenu, c.HideCompletion)
+	}
+	holder := fyne.CurrentApp().Driver().CanvasForObject(c)
+	pos := fyne.CurrentApp().Driver().AbsolutePositionForObject(c)
+
+	if c.popupMenu == nil {
+		c.popupMenu = widget.NewPopUp(c.navigableList, holder)
+	}
+	max := fyne.Min(10, float32(len(c.items)))
+	c.popupMenu.Resize(fyne.NewSize(c.Entry.Size().Width, c.popupMenu.MinSize().Height*max-3*theme.Padding()))
+	c.popupMenu.ShowAtPosition(fyne.Position{X: pos.X, Y: pos.Y + c.Size().Height})
+	holder.Focus(c.navigableList)
+}
+
+// HideCompletion hides the completion menu.
+func (c *CompletionEntry) HideCompletion() {
+	if c.popupMenu != nil {
+		c.popupMenu.Hide()
+	}
+}
+
+// SetOptions set the completion list with itemList.
+func (c *CompletionEntry) SetOptions(itemList []string) {
+	c.items = itemList
+	if c.navigableList != nil {
+		c.navigableList.SetOptions(itemList)
+	}
+}
+
+// Prevent the menu to open when the user validate value from the menu.
+func (c *CompletionEntry) setTextFromMenu(s string) {
+	c.pause = true
+	c.Entry.SetText(s)
+	c.Entry.Refresh()
+	c.pause = false
+	c.popupMenu.Hide()
+}
+
+type navigableList struct {
+	widget.List
+	data            binding.StringList
+	entry           *widget.Entry
+	selected        int
+	setTextFromMenu func(string)
+	hide            func()
+	navigating      bool
+}
+
+func newNavigableList(items []string, entry *widget.Entry, setTextFromMenu func(string), hide func()) *navigableList {
+	n := &navigableList{entry: entry, selected: -1, setTextFromMenu: setTextFromMenu, hide: hide}
+	n.data = binding.BindStringList(&items)
```

Comment:
I think you might be better off not using bindings. There is a slight performance penalty when using it and I think the code could be a bit cleaner without it. You can see https://github.com/fyne-io/fyne/pull/2181 for some samples of how I did it. That applies to the list code below as well.

Reply:
I use binding for a good reason 😄 

The main difference between SelectEntry and CompltionEntry is that I need to change the list content while the user type, without recreating it.

At this time, there is no "easy" way to reset the List items to change it.

Maybe the example I give in README is not representative, I have another example that I need:

- when user type, I do a "completion search" on a database (or API)
- I must remove the entire elements and rewrite them
- I should **not** close and open the list, I need to change the content

This capture shows the result:

https://user-images.githubusercontent.com/100080/115305424-9268a600-a166-11eb-9983-c8ad50110cc6.mp4

Or maybe you've got a nicer solution to clean the list and reset items?